account_invoice_line_stock_move_info/models/account_invoice.py

```diff
@@ -0,0 +1,38 @@
+
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from openerp import models, fields, api, exceptions, _
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    @api.multi
+    def unlink(self):
+        for invoice in self:
+            for line in invoice.invoice_line:
```

Comment:
Puedes quitarte un nivel de for poniendo:

`for line in self.mapped('invoice_line'):`

e incluso aplanarlo del todo:

`if any([x and x != 'cancel' for x in self.mapped('invoice_line.picking_id.state')]:`